nifi-nar-bundles/nifi-gcp-bundle/nifi-gcp-processors/src/main/java/org/apache/nifi/processors/gcp/bigquery/PutBigQueryStreaming.java

```diff
@@ -191,6 +195,16 @@ public void onTrigger(ProcessContext context, ProcessSession session) throws Pro
                     lmapr.add(convertMapRecord(((MapRecord) mapr).toMap()));
                 }
                 result.put(key, lmapr);
+            } else if (obj instanceof Timestamp) {
+                LocalDateTime dateTime = LocalDateTime.ofInstant(Instant.ofEpochMilli(((Timestamp) obj).getTime()), ZoneOffset.UTC);
+                DateTimeFormatter timestampFormatter = DateTimeFormatter.ofPattern("yyyy-MM-dd HH:mm:ss.SSSSSS");
+                result.put(key, dateTime.format(timestampFormatter));
+            } else if (obj instanceof Time) {
+                LocalDateTime dateTime = LocalDateTime.ofInstant(Instant.ofEpochMilli(((Time) obj).getTime()), ZoneOffset.UTC);
+                DateTimeFormatter timeFormatter = DateTimeFormatter.ofPattern("HH:mm:ss.SSSSSS");
+                result.put(key, dateTime.format(timeFormatter) );
```

Comment:
The [Time type](https://cloud.google.com/bigquery/docs/reference/standard-sql/data-types#time_type) definition indicates that the value is "independent of a specific date and timezone", so converting the value with `ZoneOffset.UTC` does not seem to match the type definition. The [java.time.LocalTime.toString()](https://docs.oracle.com/javase/8/docs/api/java/time/LocalTime.html#toString) method supports string formatting based on the precision of the original value, so it looks like this could be simplified to using `java.sql.Time.toLocalTime().toString()`.  That approach would also avoid the time zone conversion.

```suggestion
                final LocalTime localTime = ((Time) obj).toLocalTime();
                result.put(key, localTime.toString());
```

Reply:
Hi, I have already tried this method but it is dependent on the timezone in which the code is running for my part instead of having 12:35:24, it shows me 14:35:24.

Reply:
Thanks for the reply, I understand the problem. The `JsonTreeReader` and other Record Reader implementations perform implicit conversion from the local system time zone to the GMT time zone when parsing a string to a Timestamp or Time object. This unfortunately impacts a number of processors, and as you have found in this class, it ends up requiring setting `ZoneOffset.UTC` when converting back from a Timestamp or Time to String.

I addressed a similar problem with Date field handling in #5210, so that's why `java.sql.Date.toString()` works in this class.

Addressing the larger issue with Time and Timestamp in Record Readers needs to be addressed elsewhere, so for that reason, maintaining the approach you have implemented seems reasonable for now.

With that background, I would recommend keeping the `DateTimeFormatter`, but declaring it as a static class variable.  Unlike `java.text.SimpleDateFormat`, `java.time.DateTimeFormatter` is thread-safe, and can be reused, so making it a static class variable would avoid unnecessary object creation when handling each record.

It might also be helpful to include a short comment that the ZoneOffset.UTC time zone is necessary due to implicit time zone conversion in Record Readers. At least that will be a pointer to address it when evaluating the underlying issue.

Reply:
Done, thanks for the explanation and the advice.